File: ts/src/duration.ts
/**
 * Go-compatible duration formatting and parsing. The spec's display form for
 * durations is Go's `time.Duration.String()` ("5m0s", "1.5s", "300ms") and
 * string inputs accept Go's `time.ParseDuration` syntax ("2h45m", "-1.5s").
 */

import { create } from "@bufbuild/protobuf";
import type { Duration, Timestamp } from "@bufbuild/protobuf/wkt";
import { DurationSchema, TimestampSchema } from "@bufbuild/protobuf/wkt";

const NANOS = 1_000_000_000n;

/** Total nanoseconds of a wkt Duration. */
export function durationNanos(d: Duration): bigint {
  return d.seconds * NANOS + BigInt(d.nanos);
}

/** Builds a wkt Duration from total nanoseconds. */
export function durationFromNanos(total: bigint): Duration {
  let seconds = total / NANOS;
  let nanos = total % NANOS;
  if (nanos < 0n && seconds > 0n) {
    seconds -= 1n;
    nanos += NANOS;
  }
  return create(DurationSchema, { seconds, nanos: Number(nanos) });
}

/** Total nanoseconds of a wkt Timestamp (epoch-relative). */
export function timestampNanos(t: Timestamp): bigint {
  return t.seconds * NANOS + BigInt(t.nanos);
}

/** Builds a wkt Timestamp from epoch milliseconds. */
export function timestampFromMillis(ms: number): Timestamp {
  const seconds = BigInt(Math.floor(ms / 1000));
  const nanos = Math.round((ms - Number(seconds) * 1000) * 1e6);
  return create(TimestampSchema, { seconds, nanos });
}

/**
 * Formats a duration exactly like Go's `time.Duration.String()`.
 */
export function formatGoDuration(d: Duration): string {
  let v = durationNanos(d);
  const neg = v < 0n;
  if (neg) {
    v = -v;
  }

  if (v === 0n) {
    return "0s";
  }

  let out: string;
  if (v < NANOS) {
    // Sub-second: ns, µs or ms with fraction.
    if (v < 1000n) {
      out = `${v}ns`;
    } else if (v < 1_000_000n) {
      out = `${frac(v, 1000n)}µs`;
    } else {
      out = `${frac(v, 1_000_000n)}ms`;
    }
  } else {
    const minuteNs = 60_000_000_000n;
    let s = `${frac(v % minuteNs, NANOS)}s`;
    let rest = v / minuteNs; // whole minutes
    if (rest > 0n) {
      s = `${rest % 60n}m${s}`;
      rest /= 60n;
      if (rest > 0n) {
        s = `${rest}h${s}`;
      }
    }
    out = s;
  }
  return neg ? `-${out}` : out;
}

/** value/unit with trailing-zero-trimmed fraction (Go's fmtFrac). */
function frac(v: bigint, unit: bigint): string {
  const whole = v / unit;
  let fraction = (v % unit).toString().padStart(unit.toString().length - 1, "0");
  fraction = fraction.replace(/0+$/, "");
  return fraction === "" ? `${whole}` : `${whole}.${fraction}`;
}

const UNIT_NANOS = new Map<string, bigint>([
  ["ns", 1n],
  ["us", 1000n],
  ["µs", 1000n],
  ["μs", 1000n],
  ["ms", 1_000_000n],
  ["s", 1_000_000_000n],
  ["m", 60_000_000_000n],
  ["h", 3_600_000_000_000n],
]);

const DURATION_RE = /^([+-])?((\d+(\.\d*)?|\.\d+)(ns|us|µs|μs|ms|s|m|h))+$/;
const PART_RE = /(\d+(?:\.\d*)?|\.\d+)(ns|us|µs|μs|ms|s|m|h)/g;

/**
 * Parses Go `time.ParseDuration` syntax. Returns undefined on invalid input.
 */
export function parseGoDuration(s: string): Duration | undefined {
  if (s === "0") {
    return create(DurationSchema);
  }
  if (!DURATION_RE.test(s)) {
    return undefined;
  }
  const neg = s.startsWith("-");
  let total = 0n;
  for (const m of s.matchAll(PART_RE)) {
    const num = m[1];
    const unit = m[2];
    if (num === undefined || unit === undefined) {
      return undefined;
    }
    const unitNs = UNIT_NANOS.get(unit);
    if (unitNs === undefined) {
      return undefined;
    }
    const [whole = "0", fracDigits = ""] = num.split(".");
    total += BigInt(whole) * unitNs;
    if (fracDigits !== "") {
      // Fractional part scaled into the unit without float error.
      total += (BigInt(fracDigits) * unitNs) / 10n ** BigInt(fracDigits.length);
    }
  }
  return durationFromNanos(neg ? -total : total);
}

/**
 * Formats a timestamp exactly like Go's `t.Format(time.RFC3339)` in UTC:
 * whole seconds, no fractional part, trailing Z.
 */
export function formatRfc3339(t: Timestamp): string {
  const date = new Date(Number(t.seconds) * 1000);
  return `${date.toISOString().slice(0, 19)}Z`;
}

const RFC3339_RE =
  /^(\d{4})-(\d{2})-(\d{2})[Tt](\d{2}):(\d{2}):(\d{2})(\.\d+)?([Zz]|[+-]\d{2}:\d{2})$/;

/** Parses an RFC3339 string into a Timestamp. Undefined on invalid input. */
export function parseRfc3339(s: string): Timestamp | undefined {
  if (!RFC3339_RE.test(s)) {
    return undefined;
  }
  const ms = Date.parse(s);
  if (Number.isNaN(ms)) {
    return undefined;
  }
  return timestampFromMillis(ms);
}

File: rust/src/duration.rs
//! Go-compatible duration formatting/parsing and RFC3339 timestamps.

use pbjson_types::{Duration, Timestamp};

const NANOS: i128 = 1_000_000_000;

#[must_use]
pub fn duration_nanos(d: &Duration) -> i128 {
    i128::from(d.seconds) * NANOS + i128::from(d.nanos)
}

#[must_use]
#[allow(clippy::cast_possible_truncation)]
pub const fn duration_from_nanos(total: i128) -> Duration {
    let mut seconds = total / NANOS;
    let mut nanos = total % NANOS;
    if nanos < 0 && seconds > 0 {
        seconds -= 1;
        nanos += NANOS;
    }
    Duration {
        seconds: seconds as i64,
        nanos: nanos as i32,
    }
}

#[must_use]
pub fn timestamp_nanos(t: &Timestamp) -> i128 {
    i128::from(t.seconds) * NANOS + i128::from(t.nanos)
}

/// Formats a duration exactly like Go's `time.Duration.String()`.
#[must_use]
pub fn format_go_duration(d: &Duration) -> String {
    let mut v = duration_nanos(d);
    let neg = v < 0;
    if neg {
        v = -v;
    }
    if v == 0 {
        return "0s".to_owned();
    }
    let out = if v < NANOS {
        if v < 1_000 {
            format!("{v}ns")
        } else if v < 1_000_000 {
            format!("{}µs", frac(v, 1_000))
        } else {
            format!("{}ms", frac(v, 1_000_000))
        }
    } else {
        let minute = 60 * NANOS;
        let mut s = format!("{}s", frac(v % minute, NANOS));
        let mut rest = v / minute;
        if rest > 0 {
            s = format!("{}m{s}", rest % 60);
            rest /= 60;
            if rest > 0 {
                s = format!("{rest}h{s}");
            }
        }
        s
    };
    if neg {
        format!("-{out}")
    } else {
        out
    }
}

fn frac(v: i128, unit: i128) -> String {
    let whole = v / unit;
    let width = unit.to_string().len() - 1;
    let fraction = format!("{:0width$}", v % unit);
    let trimmed = fraction.trim_end_matches('0');
    if trimmed.is_empty() {
        whole.to_string()
    } else {
        format!("{whole}.{trimmed}")
    }
}

/// Parses Go `time.ParseDuration` syntax. `None` on invalid input.
#[must_use]
pub fn parse_go_duration(s: &str) -> Option<Duration> {
    if s == "0" {
        return Some(Duration::default());
    }
    let (neg, mut rest) = s
        .strip_prefix('-')
        .map_or_else(|| (false, s.strip_prefix('+').unwrap_or(s)), |r| (true, r));
    if rest.is_empty() {
        return None;
    }
    let mut total: i128 = 0;
    while !rest.is_empty() {
        let num_len = rest
            .find(|c: char| !c.is_ascii_digit() && c != '.')
            .unwrap_or(rest.len());
        if num_len == 0 {
            return None;
        }
        let (num, tail) = rest.split_at(num_len);
        let unit_len = tail
            .find(|c: char| c.is_ascii_digit() || c == '.')
            .unwrap_or(tail.len());
        let (unit, next) = tail.split_at(unit_len);
        let unit_ns: i128 = match unit {
            "ns" => 1,
            "us" | "µs" | "μs" => 1_000,
            "ms" => 1_000_000,
            "s" => NANOS,
            "m" => 60 * NANOS,
            "h" => 3_600 * NANOS,
            _ => return None,
        };
        let (whole, frac_digits) = num.split_once('.').unwrap_or((num, ""));
        let whole_n: i128 = if whole.is_empty() {
            0
        } else {
            whole.parse().ok()?
        };
        total += whole_n * unit_ns;
        if !frac_digits.is_empty() {
            let frac_n: i128 = frac_digits.parse().ok()?;
            total += frac_n * unit_ns / 10_i128.pow(u32::try_from(frac_digits.len()).ok()?);
        }
        rest = next;
    }
    Some(duration_from_nanos(if neg { -total } else { total }))
}

/// Formats a timestamp like Go's `t.Format(time.RFC3339)` in UTC: whole
/// seconds, trailing Z.
#[must_use]
pub fn format_rfc3339(t: &Timestamp) -> String {
    chrono::DateTime::from_timestamp(t.seconds, 0).map_or_else(String::new, |dt| {
        dt.format("%Y-%m-%dT%H:%M:%SZ").to_string()
    })
}

/// Parses an RFC3339 string into a Timestamp. `None` on invalid input.
#[must_use]
pub fn parse_rfc3339(s: &str) -> Option<Timestamp> {
    let dt = chrono::DateTime::parse_from_rfc3339(s).ok()?;
    Some(Timestamp {
        seconds: dt.timestamp(),
        nanos: i32::try_from(dt.timestamp_subsec_nanos()).unwrap_or(0),
    })
}
